roughpy_jax/src/cpu/dense_ft_fma.cpp

```diff
@@ -0,0 +1,193 @@
+#include "cpu/dense_ft_fma.hpp"
+
+#include <cstring>
+#include <type_traits>
+
+#include "roughpy_compute/dense/basic/free_tensor_fma.hpp"
+
+namespace {
+
+namespace ffi = xla::ffi;
+
+// FIXME split into 32 and 64 variants with templates?
+using RpyFloatType = float;
+inline constexpr ffi::DataType XlaIndexType = ffi::DataType::S32;
+inline constexpr ffi::DataType XlaFloatType = ffi::DataType::F32;
+static_assert(
+    std::is_same_v<
+        RpyFloatType,
+        ffi::NativeType<XlaFloatType>
+    >,
+    "XlaFloatType must match underlying float type"
+);
+
+} // namespace
+
+namespace rpy::jax::cpu {
+
+template <ffi::DataType T>
+std::pair<int64_t, int64_t> GetDims(const ffi::Buffer<T>& buffer)
+{
+    auto dims = buffer.dimensions();
+    if (dims.size() == 0) {
+        return std::make_pair(0, 0);
+    }
+    return std::make_pair(buffer.element_count(), dims.back());
+}
+
+// FIXME This reworked/duplicate from roughpy/compute/_src/call_config.cpp, can
+// We reuse private CallConfig to centralise this behaviour?
+std::optional<ffi::Error> update_algebra_params(
```

Comment:
The `ffi::Error` includes a "succces" option which should be the default return for a successfully completed operation.